skimage/registration/_affine.py

```diff
@@ -0,0 +1,201 @@
+        ``ndim``-dimensional space.
+    """
+    m = parameter_vector.shape[0]
+    ndim = int((np.sqrt(4*m + 1) - 1) / 2)
+    top_matrix = np.reshape(parameter_vector, (ndim, ndim+1))
+    bottom_row = np.array([[0] * ndim + [1]])
+    return np.concatenate((top_matrix, bottom_row), axis=0)
+
+
+def cost_nmi(image0, image1, *, bins=100):
+    """Negative of the normalized mutual information.
+
+    See :func:`skimage.metrics.normalized_mutual_information` for more info.
+
+    Parameters
+    ----------
+    image0, image1 : array
+        The images to be compared. They should have the same shape.
+    bins : int or sequence of int, optional
+        The granularity of the histogram with which to compare the images.
+        If it's a sequence, each number is the number of bins for that image.
+
+    Returns
+    -------
+    cnmi : float
+        The negative of the normalized mutual information between ``image0``
+        and ``image1``.
+    """
+    return -normalized_mutual_information(image0, image1, bins=bins)
+
+
+def _param_cost(reference_image, moving_image, parameter_vector, *,
+                vector_to_matrix, cost, multichannel):
+    transformation = vector_to_matrix(parameter_vector)
+    if not multichannel:
+        transformed = ndi.affine_transform(moving_image, transformation,
+                                           order=1)
+    else:
+        transformed = np.zeros_like(moving_image)
+        for ch in range(moving_image.shape[-1]):
+            ndi.affine_transform(moving_image[..., ch], transformation,
+                                 order=1, output=transformed[..., ch])
+    return cost(reference_image, transformed)
+
+
+def affine(reference_image, moving_image, *,
+           cost=cost_nmi, initial_vector=None,
+           translation_indices=None, vector_to_matrix=None,
+           pyramid_scale=2, pyramid_minimum_size=8, multichannel=False,
+           inverse=True, level_callback=None, method='Powell',
+           **kwargs):
+    """Find a transformation matrix to register a moving image to a reference.
+
+    Parameters
+    ----------
+    reference_image : ndarray
+        A reference image to compare against the target.
+    moving_image : ndarray
+        Our target for registration. Transforming this image using the
+        returned matrix aligns it with the reference.
+    cost : function, optional
+        A cost function which takes two images and returns a score which is
+        at a minimum when images are aligned. Uses the normalized mutual
+        information by default.
+    initial_vector : array of float, optional
+        The initial vector to optimize. This vector should have the same
+        dimensionality as the transform being optimized. For example, a 2D
+        affine transform has 6 parameters. A 2D rigid transform, on the other
+        hand, only has 3 parameters.
+    translation_indices : array of int, optional
```

Comment:
This + `vector_to_matrix` feels a bit strange; why pay special attention to the translation parameters?

Reply:
@stefanv see below: in the affine transform, translation parameters are the only ones that are not scale-invariant. Therefore, they need to be rescaled at each level of the pyramid, whereas the other parameters don't.

Reply:
I've added a sentence in the docstring to this effect.

Reply:
> Conversion to vector isn't well-defined. Should a Euclidean model convert itself to the linearised affine matrix, or to the rotation, scale, and translation parameters? And in what order should those parameters be? Leaving this as a function input allows us to give the user control over this.

But we only allow for one model right now, as far as I can tell; so why expose this implementation detail at all?

> The next question is whether we want to _limit_ the API to our models. I dislike this a lot — one of our guiding principles is to support APIs based on simple data structures, ie in this case, a simple NumPy array.

Yes, NumPy arrays are what we use for data transfer.  But we can return the affine parameters, in a well defined order, to the user, and use our internal model for affine transformations in the process.

> No, the bottleneck is ndimage.affine. _Once that is moved to the GPU_, the bottleneck becomes NMI.

Interesting; I thought ndimage affine was very fast.